sklearn_numba_dpex/kmeans/kernels/compute_labels.py

```diff
@@ -0,0 +1,147 @@
+    n_windows_per_centroid = math.ceil(n_features / centroids_window_height)
+
+    centroids_window_shape = (centroids_window_height, (window_n_centroids + 1))
+
+    inf = dtype(math.inf)
+
+    @dpex.kernel
+    # fmt: off
+    def assignment(
+        X_t,                      # IN READ-ONLY   (n_features, n_samples)
+        centroids_t,              # IN READ-ONLY   (n_features, n_clusters)
+        centroids_half_l2_norm,   # IN             (n_clusters,)
+        assignments_idx,          # OUT            (n_samples,)
+    ):
+    # fmt: on
+
+        sample_idx = dpex.get_global_id(0)
+        local_work_id = dpex.get_local_id(0)
+
+        centroids_window = dpex.local.array(shape=centroids_window_shape, dtype=dtype)
+        centroids_window_half_l2_norm = dpex.local.array(
+            shape=window_n_centroids, dtype=dtype
+        )
+        dot_products = dpex.private.array(shape=window_n_centroids, dtype=dtype)
+
+        first_centroid_idx = 0
+
+        min_idx = 0
+        min_sample_pseudo_inertia = inf
+
+        window_loading_centroid_idx = local_work_id % window_n_centroids
+        window_loading_feature_offset = local_work_id // window_n_centroids
+
+        for _0 in range(n_windows_per_feature):
+            _initialize_window_of_centroids(
+                local_work_id,
+                first_centroid_idx,
+                centroids_half_l2_norm,
+                centroids_window_half_l2_norm,
+                dot_products,
+            )
+
+            loading_centroid_idx = first_centroid_idx + window_loading_centroid_idx
+
+            first_feature_idx = 0
+
+            for _1 in range(n_windows_per_centroid):
+                _load_window_of_centroids_and_features(
+                    first_feature_idx,
+                    loading_centroid_idx,
+                    window_loading_centroid_idx,
+                    window_loading_feature_offset,
+                    centroids_t,
+                    centroids_window,
+                )
+
+                dpex.barrier(dpex.CLK_LOCAL_MEM_FENCE)
+
+                _accumulate_dot_products(
+                    sample_idx,
+                    first_feature_idx,
+                    X_t,
+                    centroids_window,
+                    dot_products,
+                )
+
+                dpex.barrier(dpex.CLK_LOCAL_MEM_FENCE)
+
+                first_feature_idx += centroids_window_height
+
+            min_idx, min_sample_pseudo_inertia = _update_closest_centroid(
+                first_centroid_idx,
+                min_idx,
+                min_sample_pseudo_inertia,
+                centroids_window_half_l2_norm,
+                dot_products,
+            )
+
+            dpex.barrier(dpex.CLK_LOCAL_MEM_FENCE)
+
+            first_centroid_idx += window_n_centroids
+
+        # No update step, only store min_idx in the output array
+        if sample_idx >= n_samples:
+            return
```

Comment:
Shouldn't this be moved at before for-loops?

Reply:
ah this is a good question and actually a subtle point that would benefit from a comment, IIRC I found that an undesirable work item (such as `sample_idx >= n_samples`) still has to reach the barriers, else the kernel will deadlock.

In general kernels with such code

```python
If some_condition:
    do_sth()
    dpex.barrier(...)
else:
    maybe_do_sth_else()
```

will deadlock, and must be written like this:

```python
If some_condition:
    do_sth()
else:
    maybe_do_sth_else().
    # no barrier

dpex.barrier(...)
```

Reply:
OK, I think this better be made explicit in a comment as you suggest.